### ISSUE TITLE
TextMate.Installation.Dispose does not free the TextEditor (View) for deconstruction

### ISSUE THREAD

**Issue by ArchLeaders** [AvaloniaUI/AvaloniaEdit#362]
As is probably expected: Calling the `SetGrammer` method on a `TextMate.Installation` instance keeps the owning view from being deconstructed.

However, calling `Dispose` does not unlock it when text is already loaded in the editor.

---

## To Recreate:

- Create a new Avalonia app
- Set the output type to `Exe` to show the console
- Add the `AvaloniaEdit` and `AvaloniaEdit.TextMate` packages
- Add the `StyleInclude` to the `App.axaml`
- Create a new `UserControl` with a single `TextEditor` control (xaml) and private `TextMate.Installation` field (code-behind)
- In the constructor, set the `TextMate.Installation` field from `TextEditor.InstallTextMate` and set the grammar
- Create a public function that calls `TextMate.Installation.Dispose` on the private field
- Create a deconstructor and write something to the console to know when it runs
- In `App.OnFrameworkInitializationCompleted` set `ApplicationLifetime.MainWindow` to a new window with the content being a new instance of the custom control
- Next, cast `ApplicationLifetime.MainWindow.Content` to the custom control and call the cleanup method (can't use a variable cause that will stop it from being disposed in the function)
- Wait 1 second using `await Task.Delay(1000)` and then set `ApplicationLifetime.MainWindow.Content` to `null`
- Wait another second before calling `GC.Collect()` and `GC.WaitForPendingFinalizers()`

### Expected Result

The deconstructor will run and print to the console

### Actual Result

Nothing is printed until you comment out `SetGrammar` or set the `TextEditor.Text` to an empty string

---

A minimal sample can be found on my github - https://github.com/ArchLeaders/AvaloniaEditDisposing (check the todo list to find SetGrammar)

**Comment by danipen** [comment]
Hi thanks for the report. It needs some investigation. Probably something is not released in the dispose chain and it's indirectly keeping a reference to the TextEditor.

In a first view I see that the transformed added [here](https://github.com/AvaloniaUI/AvaloniaEdit/blob/d0917422f4d524c8e99a6f04438d5c91f790f3aa/src/AvaloniaEdit.TextMate/TextMate.cs#L114) is never removed...

Are you planning a PR for a fix? 

I can take a look but it may take some time since I'm on paternity leave now.

**Comment by ArchLeaders** [comment]
Thanks for the insight. 

If it's not to complicated I should be able to manage a PR, I'll update here if I can't figure it out though.

**Comment by danipen** [comment]
Yes. Using a memory profiler should be easy to identify who is retaining the reference.

**Comment by danipen** [comment]
@ArchLeaders any updates on this?

**Comment by ArchLeaders** [comment]
> @ArchLeaders any updates on this?

No, not really, I've been pretty busy lately. I might get time in December, but work will probably take precedence.

At any rate, I have no advancements on the issue.

**Comment by udlose** [comment]
I've investigated and fixed this leak. However, during investigation, I found several race conditions in https://github.com/udlose/TextMateSharp/blob/master/src/TextMateSharp/Model/TMModel.cs that I get bitten by in my app. I've also fixed these. Because of the dependency chain, I'm going to provide my PR to https://github.com/danipen/TextMateSharp for the race conditions first. Then I'll provide the PR here to fix this memory leak.

[Here is the link](https://github.com/danipen/TextMateSharp/issues/109) to the race condition I opened for `TextMateSharp`

https://github.com/danipen/TextMateSharp/issues/109

**Comment by udlose** [comment]
@danipen 
I'm working on fixing https://github.com/danipen/TextMateSharp/issues/109 this weekend. After that PR is posted, I'll post the PR I have for this issue.

It would be ideal if the fix for TextMateSharp was included with this fix so you only have to create a single release here. If you are ok with that, my question is one about timing...

I can either:
1. post the PR for this leak here before https://github.com/danipen/TextMateSharp/issues/109 is merged and publicly released OR
2. wait to post the PR for the leak and include the package update for TextMateSharp so that both are fixed in the next release for AvaloniaEdit?

Do you have a preference?

**Comment by danipen** [comment]
Option 2 is better 👍  

I manage TextMateSharp, so releases there are usually immediate. AvaloniaEdit releases, since they’re handled by the Avalonia team, can take a bit longer.

It’s cleaner to:

- Merge and release the fix in TextMateSharp first  
- Then open the PR here including:
  - The updated TextMateSharp package  
  - The memory leak fix  
- Ship everything together in a single AvaloniaEdit release  

That way we avoid multiple releases and keep things nicely aligned.

**Comment by udlose** [comment]
Ok, cool. Option 2 was my preference as well

**Comment by udlose** [comment]
@ArchLeaders just an update here. I completed the [race conditions fixes for TextMate](https://github.com/danipen/TextMateSharp/issues/109) and @danipen has merged it. I'm working on some memory allocation improvements in TextMate. I should have those done this week. Once those are merged and a new TextMate release is published, I'll pull the new TextMate version into AvaloniaEdit, and merge it with my memory leak fix here.

1. How have you seen this leak manifest in a PROD scenario?
   a. Increased memory over time?
   b. Thru a code inspection?
   c. did your app crash?
2. Were you able to measure the impact?
3. What version of .NET were you targeting?
4. Do you have a memory dump you can share? I wanted to see the actual memory profile of the leak.

**Comment by deapstack** [comment]
Hello,

I have an issue which I think is similar (or even identical).

I have a view of my application that shows a Text editor. At load the TextEditor is configured as follows :

```
private void ConfigureAndLoadViewModel(object? sender, EventArgs? e)
    {
        _installation = _textEditor.InstallTextMate(_registryOptions);
        _installation.SetGrammar(
            _registryOptions.GetScopeByLanguageId(_registryOptions.GetLanguageByExtension(".pas").Id));
    }
```

When I close/reopen the view, I noticed an increase of memory, that led me to look further. At each time my view was re-opened, a ~5MB increase of memory happened.

When the `_installation.Dispose()` is called at view unload, that doesn't change anything.

I don't know if that can help, I am far than being an expert in DotMemory, but ... : 

<img width="2287" height="1442" alt="Image" src="https://github.com/user-attachments/assets/cdcdb23b-44f0-44f2-9cde-adc5fc47c0ca" />

I can also provide additional information if that can help...

**Comment by udlose** [comment]
Thanks for the info. I am almost done fixing some things on TextMateSharp repo. After that, @danipen  how do you want to coordinate this fix and release?

**Comment by danipen** [comment]
Once we finish all the work on the TextmateSharp repo I'll push a PR here to bump the TextmateSharp version and will merge into main. 

Then the leak can be investigated/fixed.

**Comment by udlose** [comment]
@danipen just to clarify, I have the leak identified and fixed already. I just haven't pushed a PR for this leak since we were making the TextMateSharp changes.

**Comment by danipen** [comment]
But do the leak changes require updating TextMateSharp? If not, feel free to push the PR here so we can move things forward.

**Comment by udlose** [comment]
> But do the leak changes require updating TextMateSharp? If not, feel free to push the PR here so we can move things forward.

They do not. Let me create the PR now

**Comment by udlose** [comment]
@danipen while double-checking the changes, I noticed a design issue that creates a _potential_ for a leak: the static _exceptionHandler creates a potential leak if any user code registers a handler that captures an `Installation` instance.

ex:
```
var installation = editor.InstallTextMate(options);

// LEAK: the static handler now roots 'installation' forever since it's static
TextMate.RegisterExceptionHandler(ex => 
{
    Console.WriteLine($"Error in {installation.RegistryOptions}");
    // Captures 'installation' in closure
});

installation.Dispose(); // Doesn't matter - still rooted by static field
```

how would you like to handle this? separately from this issue?

**Comment by udlose** [comment]
One other problem with the `_exceptionHandler` being static is that code reads `_exceptionHandler` at construction time to pass into child objects. That means even if someone sets the static after constructing the Installation, the child objects (TextEditorModel, TextMateColoringTransformer) wouldn't see it.

I'd recommend moving the `_exceptionHandler` field to be instance-based.

1. Add an `Action<Exception>` exceptionHandler parameter to both `InstallTextMate()` and the Installation constructor (optional, defaulting to null)
2. Store it as an instance field `_exceptionHandler` on `Installation` class
3. Remove the static `RegisterExceptionHandler()` method and the static `_exceptionHandler` field entirely

Pros:
- fixes the potential for a memory leak
- fixes the issue i mention above with order of when the static is set (before or after Installation construction)

Cons: 
- breaking API change

I can't think of another reasonable solution that solves the leak, and the inconsistency/ordering issue. Let me know how to proceed.

**Comment by udlose** [comment]
@deapstack do you use `RegisterExceptionHandler()`?

**Comment by danipen** [comment]
I think the breaking change is justified

**Comment by udlose** [comment]
are you cool if I include the fix in this PR? or do you want it as a separate one?

**Comment by danipen** [comment]
For me is ok.

**Comment by deapstack** [comment]
> [@deapstack](https://github.com/deapstack) do you use `RegisterExceptionHandler()`?

No, I don't use it currently

**Comment by udlose** [comment]
PR: https://github.com/AvaloniaUI/AvaloniaEdit/pull/560

**Comment by danipen** [comment]
@ArchLeaders @deapstack could you please confirm if https://github.com/AvaloniaUI/AvaloniaEdit/pull/560 fixes the issues reported above? Thanks!

**Comment by ArchLeaders** [comment]
Seems to be fixed for me, though it's also fixed in 11.4.1, so my issue may have been something else.

**Comment by udlose** [comment]
Does the memory increase and never get cleaned up? Remember that `Dispose`
is not deterministic and doesn't immediately reclaim memory. The GC decides
when do free the resources.

Make sure to test your app to get a few GC's. If you never see the cleanup
of the Installation instance, then it needs to be investigated again.

**Comment by udlose** [comment]
@danipen  did you ever release the version of TextMate that has the memory
leak fixes? if so, is that version updated in AvaloniaEdit?

**Comment by deapstack** [comment]
Hi @udlose 
Sorry I think that I have answered too fast. In the meantime I have deleted my message.

I still need to do some testing but actually I think that the issue is not due to TextMate.

**Comment by udlose** [comment]
lol - i was wondering where the message went 😅

lmk what u find
